crates/common/src/identity.rs

```diff
@@ -1,29 +1,116 @@
+use std::fmt;
+
 use crate::{
     prov::AgentId,
     signing::{DirectoryStoredKeys, SignerError},
 };
 
 use k256::ecdsa::{signature::Signer, Signature, SigningKey, VerifyingKey};
 use thiserror::Error;
+use tracing::warn;
 
 #[derive(Error, Debug)]
 pub enum IdentityError {
-    #[error("Invalid key store directory {0}")]
+    #[error("Failed to update context")]
+    OpaContext,
+
+    #[error("Failed to get agent id from JWT claims")]
+    JwtClaims,
+
+    #[error("Invalid key store directory: {0}")]
     KeyStore(#[from] SignerError),
-    #[error("Malformed JSON {0}")]
+
+    #[error("Malformed JSON: {0}")]
     SerdeJson(#[from] serde_json::Error),
-    #[error("Serialization error {0}")]
+
+    #[error("Serialization error: {0}")]
     SerdeJsonSerialize(String),
-    #[error("Signing error {0}")]
+
+    #[error("Signing error: {0}")]
     Signing(#[from] k256::ecdsa::Error),
 }
 
+/// Claims from a JWT, referenced in creating an AgentId for a Chronicle user
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct JwtClaims(pub serde_json::Map<String, serde_json::Value>);
+
+/// Contextual data for OPA from a request via GraphQL
+pub struct GraphQlOpaContext {
+    parent_type: serde_json::Value,
+    resolve_path: serde_json::Value,
+    identity_claims: serde_json::Value,
+}
+
+impl GraphQlOpaContext {
+    pub fn new(
+        parent_type: serde_json::Value,
+        resolve_path: serde_json::Value,
+        identity_claims: serde_json::Value,
+    ) -> Self {
+        Self {
+            parent_type,
+            resolve_path,
+            identity_claims,
+        }
+    }
+}
+
+/// Contextual data for OPA in the Transaction Processor
+#[derive(Debug)]
+pub struct TpOpaContext {
+    identity: serde_json::Value,
+    operation: serde_json::Value,
+    state: Vec<serde_json::Value>,
+}
+
+impl TpOpaContext {
+    pub fn new(
+        identity: serde_json::Value,
+        operation: serde_json::Value,
+        state: Vec<serde_json::Value>,
+    ) -> Self {
+        Self {
+            identity,
+            operation,
+            state,
+        }
+    }
+}
+
+/// Contains the scalar ID and OPA contextual data for a user established via JWT
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub struct JwtId {
+    pub id: AgentId,
+    #[serde(flatten)]
+    pub context: serde_json::Value,
```

Comment:
This would be better called claims, I think this is how we got confused

Reply:
Neither anonymous or chronicle need an inner type, they should just be standalone